clippy_lints/src/vec.rs

```diff
@@ -94,60 +84,42 @@ impl<'tcx> LateLintPass<'tcx> for UselessVec {
                 })
                 .is_continue();
 
-                let span = expr.span.ctxt().outer_expn_data().call_site;
+                let span = callsite_or_original(expr.span);
                 if only_slice_uses {
                     self.check_vec_macro(cx, &vec_args, span, expr.hir_id, SuggestedType::Array);
                 } else {
                     self.span_to_lint_map.insert(span, None);
                 }
-            }
+            },
             // if the local pattern has a specified type, do not lint.
-            else if let Some(_) = higher::VecArgs::hir(cx, expr)
-                && let Node::Local(local) = cx.tcx.hir().get_parent(expr.hir_id)
-                && local.ty.is_some()
-            {
+            Node::Local(Local { ty: Some(_), .. }) if higher::VecArgs::hir(cx, expr).is_some() => {
                 let span = expr.span.ctxt().outer_expn_data().call_site;
                 self.span_to_lint_map.insert(span, None);
-            }
+            },
             // search for `for _ in vec![...]`
-            else if let Some(parent) = get_parent_expr(cx, expr)
-                && parent.span.is_desugaring(DesugaringKind::ForLoop)
-                && self.msrv.meets(msrvs::ARRAY_INTO_ITERATOR)
+            Node::Expr(Expr { span, .. })
+                if span.is_desugaring(DesugaringKind::ForLoop) && self.msrv.meets(msrvs::ARRAY_INTO_ITERATOR) =>
             {
-                // report the error around the `vec!` not inside `<std macros>:`
-                let span = expr.span.ctxt().outer_expn_data().call_site;
```

Comment:
I guess the **real** question is: why does this return an empty span?

Reply:
in the example from the issue:

```rust
for elt in &(vec![...]) {
//         ^^^^^^^^^^^^ expr
```

`expr` is the reference to the expansion of the vec macro because `vec_args` is constructed from `higher::VecArgs::hir(cx, expr.peel_borrows())`, I think a clearer solution would be get the parent callsite of `expr.peel_borrows()`, can stick it at the top once since `outer_expn_data().call_site` seems to repeated many times currently